server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -0,0 +1,581 @@
+                if (targetColumns.get(i).column().equals(root)) {
+                    rootIdx = i;
+                    break;
+                }
+            }
+            final int rootIndex = rootIdx;
+            Function<IndexItem, Object> getValue = item -> {
+                Object val = item.insertValues()[rootIndex];
+                if (val instanceof Map<?, ?> m) {
+                    List<String> path = column.path();
+                    val = Maps.getByPath((Map<String, Object>) m, path);
+                }
+                if (val == null) {
+                    Symbol defaultExpression = ref.defaultExpression();
+                    if (defaultExpression != null) {
+                        val = defaultExpression.accept(symbolEval, Row.EMPTY).value();
+                    }
+                }
+                return val;
+            };
+            return NestableCollectExpression.forFunction(getValue);
+        }
+    }
+
+    /**
+     * For DEFAULT expressions or GENERATED columns
+     **/
+    record Synthetic(Input<?> input, ValueIndexer<Object> indexer) {
+    }
+
+    interface ColumnConstraint {
+
+        void verify(Object providedValue);
+    }
+
+    interface TableConstraint {
+
+        void verify();
+    }
+
+    record NotNullTableConstraint(ColumnIdent column, Input<?> input) implements TableConstraint {
+
+        @Override
+        public void verify() {
+            Object value = input.value();
+            if (value == null) {
+                throw new IllegalArgumentException("\"" + column + "\" must not be null");
+            }
+        }
+    }
+
+    record TableCheckConstraint(
+            Input<?> input,
+            CheckConstraint<Symbol> checkConstraint) implements TableConstraint {
+
+        @Override
+        public void verify() {
+            Object value = input.value();
+            // SQL semantics: If a column is omitted from an INSERT/UPDATE statement,
+            // CHECK constraints should not fail. Same for writing explicit `null` values.
+            if (value instanceof Boolean bool) {
+                if (!bool) {
+                    throw new IllegalArgumentException(String.format(
+                        Locale.ENGLISH,
+                        "Failed CONSTRAINT %s CHECK (%s)",
+                        checkConstraint.name(),
+                        checkConstraint.expressionStr()));
+                }
+            }
+        }
+    }
+
+    record MultiCheck(List<ColumnConstraint> checks) implements ColumnConstraint {
+
+        @Override
+        public void verify(Object providedValue) {
+            for (var check : checks) {
+                check.verify(providedValue);
+            }
+        }
+
+        public static ColumnConstraint merge(ColumnConstraint check1, ColumnConstraint check2) {
+            if (check1 instanceof MultiCheck multiCheck) {
+                multiCheck.checks.add(check2);
+                return check1;
+            }
```

Comment:
could it be other way around:
check1 - NOT a MultiCheck and check2 is a MultiCheck?
in this case we would need to operate other way around?

Reply:
No, that shouldn't happen because the MultiCheck is only used via the map.merge